### PR TITLE
Create secure API to associate emails to a specific form.

### LINKED ISSUE #491: Create secure API to associate emails to a specific form.
As an admin user I would like to associate specific emails to the form template who's owners are allowed to access the form response data.

See [design doc 003](https://docs.google.com/document/d/10cM-GocFqGW8TIjohzSNpz9m9r-CxDh6b7dtvB7dya0/edit?usp=sharing) for specific details 

Acceptance Criteria 

- [x] An API route in NextJS that accepts POST requests 
- [x] This route should sit in the following directory `pages/api/id/[form]/owners.js`
- [x] The request should have a valid user session from next auth otherwise a 403 error should be returned
- [x] If the request has a valid user session as well as  a valid government email and a valid form ID in the parameter of the request then a record should be created for them in the form_users table tied to the template ID. 
- [x] If the email is not unique for the specific form a 400 status code should be returned (use compound unique constraint on template FK and email )
- [x] If the request has a form ID that doesn't exist a 404 status code should be returned. 
- [x] If the request does not have a valid government email a 400 status code should be returned
- [x] Adequate unit testing on all levels

### PR DESCRIPTION
**Pull request by falila**
# Summary | Résumé
As an admin user I would like to associate specific emails to the form template who's owners are allowed to access the form response data.

closes #491.

# Test instructions | Instructions pour tester la modification

- Open postman UI
- Restart and run your migration
- Paste http://localhost:3000/api/id/{firmID}/owners in Postman  (POST)
- Paylaod sample: `{  "email": "gc-forms@canada.ca"  }`
- Get the cookie object from Chrome devTools and add it to Postman header.
- Hit the url (POST)  and it should return  `{    "success": {    "id": 8    }}`

 
Screen Shot 2021-12-03 at 5 32 05 PM
                                    
 
# Pull Request Checklist

- [x] Have you completely tested the functionality of change introduced in this PR? Is the PR solving the problem it's meant to solve within the scope of the related issue?
- [x] The PR does not introduce any new issues such as failed tests, console warnings or new bugs.
- [x] If this PR adds a package have you ensured its licensed correctly and does not add additional security issues?
- [x] Is the code clean, readable and maintainable? Is it easy to understand and comprehend.
- [x] Does your code have adequate comprehensible comments? Do new functions have [docstrings](https://tsdoc.org/)?
- [x] Have you modified the change log and updated any relevant documentation?
- [x] Is there adequate test coverage? Both unit tests and end-to-end tests where applicable?
- [x] If your PR is touching any UI is it accessible? Have you tested it with a screen reader? Have you tested it with automated testing tools such as axe?

### CODE REVIEW

File: lib/tsUtils.tsx [hunk 1]
```diff
@@ -8,3 +8,23 @@ export function hasOwnProperty<X extends Record<string, unknown>, Y extends Prop
...
 };
+
+/**
+ * This function checks if a given email is a government valid email.
+ * And it returns true if the email is a valid GC email otherwise false.
+ * @param email A valid government email
+ * @param domains The list of GC domains
+ * @returns {boolean} The validation result
+ */
+export const isValidGovEmail = (email: string, domains: string[]): boolean => {
```
Comment (Moro-Code) [inline lib/tsUtils.tsx:19]: Not sure this file is the appropriate place for this function. Perhaps this is better placed in the validation.tsx

File: pages/api/id/[form]/owners.ts [hunk 2]
```diff
@@ -74,4 +77,56 @@ export async function activateOrDeactivateFormOwners(
   //A 404 status code for a form Not Found in form_users
   return res.status(404).json({ error: "Form or email Not Found" });
 }
+/**
+ * This method aims to create a unique binding between a form template and a user
+ * who needs to access the form's data later on. Here are some requirements that must be satisfied
+ * in order to associate users to collect a form's data
+ * Rule 1: a valid government email address
```
Comment (Moro-Code) [inline pages/api/id/[form]/owners.ts:84]: Good description

File: pages/api/id/[form]/owners.ts [hunk 3]
```diff
@@ -74,4 +77,56 @@ export async function activateOrDeactivateFormOwners(
...
+  if (!requestBody?.email || !isValidGovEmail(requestBody.email, emailDomainList.domains)) {
+    return res.status(400).json({ error: "The email is not a valid GC email" });
+  }
+  const formID = req.query.form as string;
+  if (!formID) return res.status(400).json({ error: "Malformed API Request Invalid formID" });
+  //Checking if formID exists in db return true or false
+  //May return true if it exists otherwise false
+  const isFormIDExistResult = await executeQuery(
+    await dbConnector(),
+    "SELECT exists(SELECT 1 FROM templates WHERE id = ($1))",
```
Comment (Moro-Code) [inline pages/api/id/[form]/owners.ts:103]: Hmmm you don't need this nested query or this query at all infact. It's why we set a unique constraint on the template id and the email :). If you try to create a record and it fails with a unique constraint error then you know the record with the template id and the email already exists. There is also a NOT NULL constraint so if that fails you know the form doesn't exist. In this way we don't need to fire off extra queries and load the database more

File: pages/api/id/[form]/owners.ts [hunk 4]
```diff
@@ -74,4 +77,56 @@ export async function activateOrDeactivateFormOwners(
...
+  );
+  type ARowExists = { exists?: boolean };
+  const { exists } = isFormIDExistResult.rows[0] as ARowExists;
+  if (!exists) return res.status(404).json({ error: "FormID does not exist" });
+
+  const { email } = requestBody;
+  //May return true if the email is associated otherwise false
+  const emailTieToFormResult = await executeQuery(
+    await dbConnector(),
+    "SELECT exists(SELECT 1 FROM form_users WHERE template_id = ($1) AND email = ($2))",
```
Comment (Moro-Code) [inline pages/api/id/[form]/owners.ts:114]: Please see above comment

**Review by Moro-Code — CHANGES_REQUESTED** [review]
Changes needed but good work !

File: lib/tsUtils.tsx (hunk 1 above)
Comment (falila) [reply lib/tsUtils.tsx:19]: Yeah i can be moved in the validation.tsx file as well.

File: pages/api/id/[form]/owners.ts [hunk 5]
```diff
@@ -55,12 +58,16 @@ export async function activateOrDeactivateFormOwners(
 ): Promise<void> {
   //Extracting req body
   const requestBody = req.body ? JSON.parse(req.body) : undefined;
-  //Payload validation
-  if (!requestBody?.email || !requestBody?.active || typeof requestBody.active !== "boolean") {
+  //Payload validation fix: true case scenario
+  if (
+    !requestBody?.email ||
+    requestBody.active === undefined ||
```
Comment (Moro-Code) [inline pages/api/id/[form]/owners.ts:64]: you could probably just remove this statement all together. if active is undefined typeof will yield undefined

**Review by Moro-Code — CHANGES_REQUESTED** [review]
Minor changes needed

File: pages/api/id/[form]/owners.ts [hunk 6]
```diff
@@ -74,4 +81,41 @@ export async function activateOrDeactivateFormOwners(
...
+  if (!formID) return res.status(400).json({ error: "Malformed API Request Invalid formID" });
+  try {
+    const result = await executeQuery(
+      await dbConnector(),
+      "INSERT INTO form_users (template_id, email) VALUES ($1, $2) RETURNING id",
+      [formID, requestBody.email as string]
+    );
+    return res.status(200).json({ success: result.rows[0] });
+  } catch (error) {
+    const message = error.message as unknown as string;
```
[the remark this answers is not in the document]
Comment (falila) [reply pages/api/id/[form]/owners.ts:111]: TSlint/ESlint issue.

File: lib/tests/validation.test.js [hunk 7]
```diff
@@ -359,3 +359,32 @@ describe("Test getErrorList function", () => {
     spy.mockRestore();
   });
 });
+
+import { isValidGovEmail } from "@lib/validation";
```
[the remark this answers is not in the document]
Comment (falila) [reply lib/tests/validation.test.js:363]: a small mistake :)

**Review by Moro-Code — APPROVED** [review]
Looking good !

**Pull request merged**